mongodbatlas/resource_mongodbatlas_cluster.go

```diff
@@ -1361,22 +1376,188 @@ func flattenReplicationSpecs(rSpecs []matlas.ReplicationSpec) []map[string]inter
 		specs = append(specs, spec)
 	}
 
-	return specs
+	return updateRegionNames(d, specs)
+}
+
+// User can use either Atlas region names or region names of their cloud provider. Atlas cluster APIs use only Atlas regions,
+// so we update region_name usages in replication_specs to what the user has in their Terraform config
+func updateRegionNames(d *schema.ResourceData, actualSpecs []map[string]interface{}) []map[string]interface{} {
+	var providerName, backingProviderName string
+
+	if _, ok := d.GetOk("replication_specs"); !ok {
+		return actualSpecs
+	}
+
+	if val, ok := d.GetOk("provider_name"); ok {
+		providerName = val.(string)
+	}
+
+	if val, ok := d.GetOk("backing_provider_name"); ok {
+		backingProviderName = val.(string)
+	}
+
+	stateReplicationSpecsMap := getStateFileSpecsMap(d.Get("replication_specs"), providerName, backingProviderName)
+
+	for _, actualSpec := range actualSpecs {
+		hash := getReplicationSpecHash(actualSpec, true)
+		matchedStateReplicationSpec := stateReplicationSpecsMap[hash]
+
+		for _, actualRegionConfigElem := range actualSpec["regions_config"].([]map[string]interface{}) {
+			actualRegionConfig := actualRegionConfigElem
+			regionConfigHash := getRegionConfigHash(actualRegionConfig)
+			matchedStateRegionConfig := matchedStateReplicationSpec[regionConfigHash].(map[string]interface{})
+
+			actualRegionConfig["region_name"] = matchedStateRegionConfig["region_name_by_user"]
+		}
+	}
+
+	return actualSpecs
+}
+
+// This method creates a map of all the replication_specs present in Terraform state for the cluster
+// keyed on a hash value generated from property values of a replication_spec(s).
+//
+// The value of each entry is another map keyed on a hash value generated from property values of each region_config(s) in the replication_spec
+//
+// A "region_name_by_user" property is also added to each region_config to keep track if the user is
+// using an Atlas region name or a Cloud Provider's region name.
+func getStateFileSpecsMap(stateSpecs interface{}, providerName, backingProviderName string) map[uint64]map[uint64]interface{} {
```

Comment:
[nit] golang comment for a function should start with the function name https://tip.golang.org/doc/comment
```suggestion
// getStateFileSpecsMap creates a map of all the replication_specs present in Terraform state for the cluster
// keyed on a hash value generated from property values of a replication_spec(s).
//
// The value of each entry is another map keyed on a hash value generated from property values of each region_config(s) in the replication_spec
//
// A "region_name_by_user" property is also added to each region_config to keep track if the user is
// using an Atlas region name or a Cloud Provider's region name.
func getStateFileSpecsMap(stateSpecs interface{}, providerName, backingProviderName string) map[uint64]map[uint64]interface{} {
```